xtask/codegen/gritql.ungram

```diff
@@ -0,0 +1,554 @@
+
+GritPatternOr = 'or' '{' patterns: GritPatternList '}'
+
+GritPatternOrElse = 'orelse' '{' patterns: GritPatternList? '}'
+
+GritPatternAny = 'any' '{' patterns: GritPatternList? '}'
+
+GritPatternAnd = 'and' '{' patterns: GritPatternList? '}'
+
+GritPatternMaybe = 'maybe' pattern: MaybeCurlyGritPattern
+
+GritPatternAfter = 'after' pattern: AnyGritPattern
+
+GritPatternBefore = 'before' pattern: AnyGritPattern
+
+GritPatternContains =
+    'contains'
+    contains: MaybeCurlyGritPattern
+    GritPatternContainsUntilClause?
+
+GritPatternContainsUntilClause =
+    'until'
+    until: AnyGritPattern
+
+GritPatternIncludes = 'includes' includes: MaybeCurlyGritPattern
+
+GritRewrite =
+    left: AnyGritPattern
+    annotation: GritAnnotation?
+    '=>'
+    right: AnyGritPattern
+
+// --- conditional pattern ------------
+GritPatternIfElse =
+    'if'
+    '('
+    if_predicate: AnyGritPredicate
+    ')'
+    then_pattern: MaybeCurlyGritPattern
+    GritPatternElseClause?
+
+GritPatternElseClause =
+    'else'
+    else_pattern: MaybeCurlyGritPattern
+// --- conditional pattern ------------
+
+GritWithin =
+    'within'
+    pattern: MaybeCurlyGritPattern
+
+GritBubbleScope = '(' GritVariableList? ')'
+
+GritBubble =
+    'bubble'
+    variables: GritBubbleScope?
+    pattern: MaybeCurlyGritPattern
+
+MaybeGritNamedArg =
+    AnyGritPattern
+    | GritNamedArg
+    | GritBogusNamedArg
+
+GritNamedArg =
+    name: GritName
+    '='
+    pattern: AnyGritPattern
+
+GritNamedArgList = MaybeGritNamedArg (',' MaybeGritNamedArg)* ','?
+
+// maybe we should have a different fieldname for each choice?
+GritNodeLike =
+    name: GritName
+    '('
+    named_args: GritNamedArgList?
+    ')'
+
+GritLike =
+    'like'
+    GritLikeThreshold?
+    '{'
+    example: AnyGritPattern
+    '}'
+
+GritLikeThreshold =
+    '('
+    threshold: AnyGritPattern
+    ')'
+
+GritMap =
+    '{'
+    elements: GritMapElementList?
+    '}'
+
+GritMapElementList = GritMapElement (',' GritMapElement)* ','?
+
+GritMapElement =
+    key: GritName
+    ':'
+    value: AnyGritPattern
+
+GritMapAccessor =
+    map: GritMapAccessorSubject
+    '.'
+    key: GritMapKey
+    GritBogus?
+
+GritMapAccessorSubject = GritMap | AnyGritContainer
+GritMapKey = GritName | GritVariable
+
+GritList =
+    GritName?
+    '['
+    patterns: GritListPatternList?
+    ']'
+
+GritListPatternList = AnyGritListPattern (',' AnyGritListPattern)* ','?
+
+AnyGritListPattern = AnyGritPattern | GritDotdotdot
+
+GritListAccessor =
+    list: GritListAccessorSubject
+    '['
+    index: GritListIndex
+    ']'
+    GritBogus?
+
+GritListAccessorSubject = GritList | AnyGritContainer
+GritListIndex = AnyGritContainer | GritNegativeIntLiteral | GritIntLiteral
+
+GritDot = '.'
+
+GritSome = 'some' pattern: MaybeCurlyGritPattern
+GritEvery = 'every' pattern: MaybeCurlyGritPattern
+
+GritDotdotdot =
+    dotdotdot: '...'
+    MaybeCurlyGritPattern?
+
+GritUnderscore =
+    dollar_underscore: '$_'
```

Comment:
It's worth adding a comment explaining why we have a node with a keyword. I think this should be reviewed .

Usually, we use this approach only for dealing with primitives:

```
CssCustomIdentifier = value: 'ident'
```

Reply:
It's because tokens cannot be used directly inside an `Any*` along with other nodes.